### ISSUE TITLE
[Enhancement]: Lightmap Export Improvements and Export Option (PNG or EXR)

### ISSUE THREAD

**Issue by Minileandro** [AssetRipper/AssetRipper#867]
### Describe the new feature or enhancement

During some conversations on the discord server, I said that the lightmap between the original game and the one exported by Asset Ripper had some differences, among them in AR it seems to have lighter shadows and a slightly reduced quality compared to the original:

How asset ripper exported:

![AR PNG](https://github.com/AssetRipper/AssetRipper/assets/53369145/3c9374d9-54f4-48b1-aedc-ce468d74e717)

How it looks in game:

![Original](https://github.com/AssetRipper/AssetRipper/assets/53369145/0b4a5f9d-25b8-4ed7-a84b-995c301f7517)

After some tests using the **RenderDoc** program to export the lightmap in png and exr, I could see that there were some changes, in shadow color, or even comparassion between png of AR & Renderdoc (both in High quality on unity)

Lightmap in EXR (using Renderdoc):

Better in quality, but worst shadow color if compare with original
![RenderDocEXR](https://github.com/AssetRipper/AssetRipper/assets/53369145/30a2d08e-b63e-404c-8000-71c286135e3c)

Lightmap in PNG (using Renderdoc):

Same result in terms of shadow color, but better quality if compare with Asset Ripper (again, both in High Quality setting)
![RenderDoc PNG](https://github.com/AssetRipper/AssetRipper/assets/53369145/ad2799a5-11bb-4e6d-aa26-65c91fddaa66)

After some investigation and several tests, I still haven't been able to come up with an exact solution for the solution, but I think one of the main reasons is about the format.
When looking at the unity documentation, we see this information:

![image](https://github.com/AssetRipper/AssetRipper/assets/53369145/e7fe923a-7264-430d-9057-68395ae15b36)
(https://docs.unity3d.com/2017.4/Documentation/Manual/Lightmaps-TechnicalInformation.html)

And this is the information in PNG (AR version):
![image](https://github.com/AssetRipper/AssetRipper/assets/53369145/0ecd5e94-d789-407c-af8d-451c2ba3e563)

This is the information in PNG (Renderdoc version)
![image](https://github.com/AssetRipper/AssetRipper/assets/53369145/142681f5-eb4d-47f3-8920-2b59ca8001ee)

Did a lightmap in a test scene, with the settings that a old dev of this game gave to me, this is how looks:
Unity did in EXR format
![image](https://github.com/AssetRipper/AssetRipper/assets/53369145/0bdacf6f-f987-4882-81c8-978dce428694)

And how looks in my test scene:
![image](https://github.com/AssetRipper/AssetRipper/assets/53369145/ca547ac3-b4fa-4a47-a5d2-9feb523cbb17)

I believe this is the problem about export in png, but even so it would be preferable to give the user a choice between PNG or EXR, and improve the export in PNG to have the same quality.

**Comment by ds5678** [comment]
It's possible that this is (at least partially) caused by AssetRipper's image encoding practices. PNG supports 16 bit channels, but AssetRipper only ever uses 8 bit channels. If the source data is greater than 8 bits, that would cause a small quality loss during export. Unity supports up to 32 bits per channel.

**Comment by Minileandro** [comment]
Makes sense, I already thought that was it, because there really is a small loss of quality.
Now about the difference between lightmap shadow colors, I imagine that EXR would solve it, but I'll investigate a little more about it.

**Comment by ds5678** [comment]
In addition, EXR supports floating point values, which PNG does not. The conversion from the floating point scale to integer scale is lossy and would also contribute to the problem.

Assuming a suitable .NET library can be found for encoding EXR, it may be superior to PNG and worthy of being the default.

**Comment by Minileandro** [comment]
So, if Asset Ripper eventually supports EXR, it would be beneficial for this case, right?


(Btw I found out yesterday one thing about cubemaps, where it would have been enough just to have support for the .DDS format to solved the problem of the others missing parts lol)

**Comment by ds5678** [comment]
DDS and KTX would also be beneficial for AssetRipper to support. As I understand, supporting them would allow AssetRipper to avoid lossy texture decompression in some situations.

**Comment by ds5678** [comment]
EXR would be beneficial for all cases, not just lightmap.

**Comment by Minileandro** [comment]
Hm, sounds interesting then...
Btw, decompiled my project to test how lightmap looks after decompiles vs original and... xD

Original:
![image](https://github.com/AssetRipper/AssetRipper/assets/53369145/6ff80cb4-981d-43f0-b8fe-4efdd5de1cb1)

Asset Ripper:
![image](https://github.com/AssetRipper/AssetRipper/assets/53369145/1387ad53-702e-46fc-97cd-e794a87d8ee4)

**Comment by ds5678** [comment]
Comparing your images, I see both the decompression artifacts and the difference in shadow depth.

About the shadow depth, I think that may be caused by https://github.com/AssetRipper/TextureDecoder/issues/19

**Comment by Minileandro** [comment]
Hm maybe, I was really thinking that lightmap be in png would be the big problem

**Comment by Minileandro** [comment]
Make sense yeah, i will wait until be fixed  on future then!
Then maybe i should close this topic, or keep open?

**Comment by ds5678** [comment]
Keep it open. EXR would be beneficial.

**Comment by Minileandro** [comment]
I think I kind of got to the problem of different shades, but still no solution.

After making my own lightmaps I noticed a difference, which is that my lightmaps generate a kind of "background", while the original lightmap is transparent, which means that when opening the project for the first time, all room looks dark.

My custom lightmap:
![image](https://github.com/AssetRipper/AssetRipper/assets/53369145/994fc311-3af1-4077-8672-4a3f41f56878)

Lightmap from game:
![Lightmap-0_comp_light](https://github.com/AssetRipper/AssetRipper/assets/53369145/cea1710d-4355-45bc-bc6b-1de7c761ace4)

Maybe is some kinda of compression? or some method to make lightmaps too? idk, been testing a little and dont found a solution about this.

**Comment by ds5678** [comment]
I have a potential solution. I imagine that the lightmap textures don't need modified. If I export them as yaml, the output should be perfect, similar to meshes.

**Comment by Minileandro** [comment]
Probably, just testing to really know if will work or no, but seems be the answer for the problem.
Also because when I open the project for the first time, the room is all dark, but the shadows seems me correct, only after modifying the texture type does everything change.

**Comment by Minileandro** [comment]
I remembered now, I think to corraborate with you, when I exported the lightmap I made, leaving it in Default or Lightmap was indifferent, really what affected it was the quality that was in normal and had to be High, where is no more a problem now, so give a option for YAML can be a good idea.

**Comment by ds5678** [comment]
@Minileandro Let me know if that commit fixes things.

**Comment by Minileandro** [comment]
Worked!!!! :D
Finally!
![image](https://github.com/AssetRipper/AssetRipper/assets/53369145/2ffb3047-2870-4a90-a6bf-f215902b38a8)

**Comment by ds5678** [comment]
You're welcome 😄

**Comment by ds5678** [comment]
I'm closing this since the primary issue we discussed has been solved. I've created a new issue to discuss the possibility of additional image formats.
